src/Database/LSMTree/Normal.hs

```diff
@@ -335,7 +337,24 @@ withCursor ::
   => TableHandle m k v blob
   -> (Cursor m k v blob -> m a)
   -> m a
-withCursor (Internal.NormalTable th) action = Internal.withCursor th (action . Internal.NormalCursor)
+withCursor (Internal.NormalTable th) action =
+    Internal.withCursor Nothing th (action . Internal.NormalCursor)
+
+{-# SPECIALISE withCursorAtOffset :: SerialiseKey k => k -> TableHandle IO k v blob -> (Cursor IO k v blob -> IO a) -> IO a #-}
+-- | A variant of 'withCursor' that allows initialising the cursor at an offset,
+-- so it will skip all entries with keys less than the provided offset key.
+--
+-- NOTE: The ordering of the serialised keys will be used, which can lead to
+-- unexpected results if the 'SerialiseKey' instance is not order-preserving!
+withCursorAtOffset ::
+     (IOLike m, SerialiseKey k)
+  => k
+  -> TableHandle m k v blob
+  -> (Cursor m k v blob -> m a)
+  -> m a
+withCursorAtOffset offset (Internal.NormalTable th) action =
+    Internal.withCursor (Just (Internal.serialiseKey offset)) th $
```

Comment:
Better to do the serialisation here rather than defer it lazily.
```suggestion
    let !offset' = Internal.serialiseKey offset in
    Internal.withCursor (Just offset') th $
```
Alternatively, if we did the thing about using a custom data type then the strictness would propagate here automatically without any extra bang patterns at the call sites.

Reply:
I went with the `OffsetKey` data type.